### DISCUSSION TITLE
Question about the Indirect Predictor Refactor

### DISCUSSION THREAD

**Discussion by OdnetninI** [gem5/gem5#954] [gem5-dev]
Hi!
I was working on the indirect predictor and I found something weird regarding the commit: "cpu: Refactor indirect predictor " ( https://github.com/gem5/gem5/commit/f65df9b959d8e2fc3b57d8204f481c2fca912b28 ).

The indirect predictor is checked for all indirect branches that are not function returns:

```C
        // The target address is not predicted by RAS.
        // Thus, BTB/IndirectBranch Predictor is employed.
        if (!inst->isReturn()) {
            if (inst->isDirectCtrl() || !iPred) {
```

After doing the prediction, the indirect predictor is used to perform a `recordIndirect` to save the prediction into the history.

https://github.com/gem5/gem5/commit/f65df9b959d8e2fc3b57d8204f481c2fca912b28#diff-6b913cefadced10d4eefae514686d1bedb4709dc4896cfef0a1b3e5fab3b5a4eL276

This functionality was moved to the new update `function`, also executed after the prediction:

Old Code: https://github.com/gem5/gem5/blob/aa949dba0f33d04a85903fa8b1096702c2ddba7e/src/cpu/pred/simple_indirect.cc#L120-L126

New Code: https://github.com/gem5/gem5/blob/7d62da6d10fe54b9a9967e4eaae9554a75990c06/src/cpu/pred/simple_indirect.cc#L240-L248

As you can see, in the new code, this update is protected under the `history->was_indirect` flag.
The flag is set to true every time a prediction is done by the indirect predictor, but it is updated in the update function ( https://github.com/gem5/gem5/blob/7d62da6d10fe54b9a9967e4eaae9554a75990c06/src/cpu/pred/simple_indirect.cc#L218C5-L218C57 ).

The function `isIndirectNoReturn` was added in the refactor commit: https://github.com/gem5/gem5/commit/f65df9b959d8e2fc3b57d8204f481c2fca912b28#diff-ed549c292364838d1a0f401121baf9b4e3ac755cde44fda070214d9a802e7ec4R161-R164

```C
    inline bool isIndirectNoReturn(BranchType type) {
        return (type == BranchType::CallIndirect) ||
               (type == BranchType::IndirectUncond);
    }
```

Is it missing the Indirect Conditional branches?

Previous to the commit, the only checks were:
- Is not a Return
- Is not a Direct Branch

In the pull request ( https://github.com/gem5/gem5/pull/429 ) there is a comment confirming this new version has the exactly same results as the previous version ( https://github.com/gem5/gem5/pull/429#issuecomment-1761280993 ).
This makes sense in x86 because there are no Indirect Conditional branches.

**Comment by giactra** [comment]
@dhschall Yes, it looks like a bug. There is no Indirect Conditional branch in Arm neither. Regarding other ISAs, in gem5 only PowerPC seems to implement them https://github.com/gem5/gem5/blob/stable/src/arch/power/isa/formats/branch.isa#L155.

Feel free to open a PR if you want to fix this
